### PR TITLE
Add PVS-Studio-tool

### PR DESCRIPTION
**Pull request by GeorgeGribkov**
Hello!
I am a developer of PVS-Studio static code analyzer. 
I have added PVS-Studio into Compiler Explorer as a tool.

### CODE REVIEW

File: lib/tooling/PVS-Studio-tool.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2018, Compiler Explorer Authors
```
Comment (partouf) [inline lib/tooling/PVS-Studio-tool.js:1]: 2020

File: etc/config/c++.amazon.properties [hunk 2]
```diff
@@ -1214,7 +1214,14 @@ libs.namedtype.versions.trunk.path=/opt/compiler-explorer/libs/NamedType
...
 
-tools=clangtidytrunk:llvm-mcatrunk:pahole:clangformattrunk:clangquerytrunk:readelf:x86to6502:ldd:iwyu
+tools=PVS-Studio:clangtidytrunk:llvm-mcatrunk:pahole:clangformattrunk:clangquerytrunk:readelf:x86to6502:ldd:iwyu
+
+tools.PVS-Studio.name=PVS-Studio
+tools.PVS-Studio.exe=/usr/bin/pvs-studio-analyzer
+tools.PVS-Studio.type=postcompilation
+tools.PVS-Studio.exclude=cl19:cl19_2015_u3:cl_new:icc:rvclang:wasmclang:cross:ellcc:zapcc:djggp:www.godbolt.ms@443:armclang32:armclang64
+tools.PVS-Studio.class=PVS-Studio-tool
+tools.clangtidytrunk.stdinHint=disabled
```
Comment (partouf) [inline etc/config/c++.amazon.properties:1224]: tools.PVS-Studio.stdinHint I assume?

File: etc/config/c.amazon.properties [hunk 3]
```diff
@@ -569,7 +569,14 @@ libs.hedley.versions.v12.path=/opt/compiler-explorer/libs/hedley/v12/
...
 
-tools=clangtidytrunk:clangformattrunk:clangquerytrunk:llvm-mcatrunk:pahole:readelf:ldd
+tools=PVS-Studio:clangtidytrunk:clangformattrunk:clangquerytrunk:llvm-mcatrunk:pahole:readelf:ldd
+
+tools.PVS-Studio.name=PVS-Studio
+tools.PVS-Studio.exe=/usr/bin/pvs-studio-analyzer
+tools.PVS-Studio.type=postcompilation
+tools.PVS-Studio.exclude=ccl19:ccl19_2015_u3:ccl_new:armcclang32:armcclang64:rvcclang:ppci:cicc:cross:cgcc-classic:cc65
+tools.PVS-Studio.class=PVS-Studio-tool
+tools.clangtidytrunk.stdinHint=disabled
```
Comment (partouf) [inline etc/config/c.amazon.properties:579]: PVS-Studio instead of clangtidytrunk

File: etc/config/c++.amazon.properties (hunk 2 above)
Comment (GeorgeGribkov) [reply etc/config/c++.amazon.properties:1224]: Yes :)

File: lib/tooling/PVS-Studio-tool.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        args.push("--cl-params");
+        args = args.concat(compileFlags);
+
+        // If you are to modify this code,
+        // don't forget that super.runTool() does args.push(inputFilepath) inside.
+        const result = await super.runTool(compilationInfo, inputFilepath, args);
+
+        // Convert log to readable format
+        let plogConverterResult;
+        let plogConverterPath = "/usr/bin/plog-converter";
```
Comment (mattgodbolt) [inline lib/tooling/PVS-Studio-tool.js:77]: Can you make this, and `plogConverterOutputFilePath` `const` please?

File: lib/tooling/PVS-Studio-tool.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        args.push("--preprocessor");
+        if (compilationInfo.compiler.group.indexOf("clang") !== -1)
+            args.push("clang");
+        else
+            args.push("gcc");
+
+        args.push("--cl-params");
+        args = args.concat(compileFlags);
+
+        // If you are to modify this code,
```
Comment (mattgodbolt) [inline lib/tooling/PVS-Studio-tool.js:71]: Thanks for the handy comment here! :)

File: lib/tooling/PVS-Studio-tool.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        plogConverterArgs.push("-t", "errorfile", "-a", "FAIL:1,2,3;GA:1,2,3", "-o",
+            plogConverterOutputFilePath,
+            outputFilePath);
+
+        plogConverterResult = await exec.execute(plogConverterPath, plogConverterArgs, this.getDefaultExecOptions());
+        let plogConverterOutput = (await fs.readFile(plogConverterOutputFilePath, 'utf8')).toString();
+
+        // Sometimes if a code fragmet can't be analyzed, PVS-Studio makes a warning for a preprocessed file
+        // (*.PVS-Studio.i)
+        // This name can't be parsed by utils.parseOutput
```
Comment (mattgodbolt) [inline lib/tooling/PVS-Studio-tool.js:90]: Thanks!

File: lib/tooling/PVS-Studio-tool.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        // so let's just replace it with a source file name.
+        plogConverterOutput = plogConverterOutput.replace(sourceDir + "/example.PVS-Studio.i", inputFilepath);
+
+        result.stdout = utils.parseOutput(plogConverterOutput, plogConverterResult.filenameTransform(inputFilepath));
+
+        // Now let's trim the documentation link
+        if(result.stdout.length > 0)
```
Comment (mattgodbolt) [inline lib/tooling/PVS-Studio-tool.js:97]: Perhaps later on we can add support to turn this into an actual link?

**Comment by mattgodbolt** [comment]
Hi! I just deployed to our beta site. I tried : https://godbolt.org/beta/z/NkRgG-

No luck - there's no output at all in the tool window. In the logfile on the server I see:

```
May 21 21:12:09 ip-172-30-2-193 amazon.beta error: Error during compilation: ENOENT: no such file or directory, open '/tmp/compiler-explorer-compiler2020422-1529-1xfedda.qh2vk/pvs-studio-log.err' {"errno":-2,"code":"ENOENT","syscall":"open","path":"/tmp/compiler-explorer-compiler2020422-1529-1xfedda.qh2vk/pvs-studio-log.err","stack":"Error: ENOENT: no such file or directory, open '/tmp/compiler-explorer-compiler2020422-1529-1xfedda.qh2vk/pvs-studio-log.err'"} 
```

I suspect there's a problem with our sandboxing technology and pvs studio, or else the version we're running isn't working for some reason. The version on the server is:

```
$ /opt/compiler-explorer/pvs-studio-latest/bin/pvs-studio --version
PVS-Studio 7.07.38727.1022
```

File: lib/tooling/PVS-Studio-tool.js (hunk 4 above)
Comment (mattgodbolt) [inline lib/tooling/PVS-Studio-tool.js:77]: Ah... what is `/usr/bin/plog-converter` ? I managed to hack to get this running locally but I don't know where to get a `plog-converter` from. Looks like it's https://github.com/viva64/plog-converter ? If so we need to make its installation path not `/usr/bin/plog-converter`. Will have a poke around.

**Comment by partouf** [comment]
Closing in favor of https://github.com/compiler-explorer/compiler-explorer/pull/2003

**Pull request closed without merging**